### PR TITLE
domain: make ru_stats duration compatible with DST

### LINKED ISSUE #52353: pkg/domain: Test case `TestGetLastExpectedTime` failed in `Australia/Lord_Howe` timezone.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Run `TestGetLastExpectedTime` in TZ `Australia/Lord_Howe`:

```
➜  tidb git:(master) ✗ TZ="Australia/Lord_Howe" go test --tags=intest -run ^TestGetLastExpectedTime$ "github.com/pingcap/tidb/pkg/domain"

--- FAIL: TestGetLastExpectedTime (0.00s)
    ru_stats_test.go:144:
        	Error Trace:	/Users/bb7133/Projects/gopath/src/github.com/pingcap/tidb/pkg/domain/ru_stats_test.go:144
        	Error:      	Not equal:
        	            	expected: time.Date(2023, time.December, 28, 10, 0, 0, 0, time.Local)
        	            	actual  : time.Date(2023, time.December, 28, 9, 0, 0, 0, time.Local)

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-(time.Time) 2023-12-28 10:00:00 +1100 +11
        	            	+(time.Time) 2023-12-28 09:00:00 +1100 +11

        	Test:       	TestGetLastExpectedTime
FAIL
FAIL	github.com/pingcap/tidb/pkg/domain	1.402s
FAIL
```

### 4. What is your TiDB version? (Required)

Latest master(555ce0235229dc370cc8e63574437cf9ee60bdd0)

### PR DESCRIPTION
**Pull request by glorv**
### What problem does this PR solve?

Issue Number: close #52353

Problem Summary:

### What changed and how does it work?
Use the time API to calculate the start_time of ru_stats so it can be compatible with DST(daylight saving time).

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/domain/ru_stats.go [hunk 1]
```diff
@@ -108,12 +108,26 @@ func (do *Domain) requestUnitsWriterLoop() {
...
+// GetLastExpectedTimeTZ return the last written ru time under specifical timezone.
+// make it public only for test.
+func GetLastExpectedTimeTZ(now time.Time, interval time.Duration, tz *time.Location) time.Time {
+	year, month, day := now.Date()
+	start := time.Date(year, month, day, 0, 0, 0, 0, tz)
+	// cast to int64 to bypass the durationcheck lint.
+	count := int64(now.Sub(start) / interval)
+	targetDur := time.Duration(count) * interval
+	// use UTC timezone to calculate target time so it can be compatible with DST.
+	return start.In(time.UTC).Add(targetDur).In(tz)
```
Comment (wjhuang2016) [inline pkg/domain/ru_stats.go:130]: Can tz be nil here?

Comment (glorv) [reply pkg/domain/ru_stats.go:133]: No, this caller should always a valid tz. Any, I added a check to set time.Local for nil tz.

**Review by wjhuang2016 — APPROVED** [review]

**Review by nolouch — APPROVED** [review]
lgtm

**Pull request merged**